diff --git a/Dockerfile b/Dockerfile
index a335112cc..471044879 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -77,11 +77,14 @@ FROM --platform=arm64 ghcr.io/siderolabs/util-linux:${PKGS} AS pkg-util-linux-ar
 FROM --platform=amd64 ghcr.io/siderolabs/kmod:${PKGS} AS pkg-kmod-amd64
 FROM --platform=arm64 ghcr.io/siderolabs/kmod:${PKGS} AS pkg-kmod-arm64
 
-FROM ghcr.io/siderolabs/kernel:${PKGS} AS pkg-kernel
+#FROM ghcr.io/siderolabs/kernel:${PKGS} AS pkg-kernel
+FROM 127.0.0.1:5000/rsmitty/kernel:v1.5.0-dirty AS pkg-kernel
 FROM --platform=amd64 ghcr.io/siderolabs/kernel:${PKGS} AS pkg-kernel-amd64
-FROM --platform=arm64 ghcr.io/siderolabs/kernel:${PKGS} AS pkg-kernel-arm64
+FROM --platform=arm64 127.0.0.1:5000/rsmitty/kernel:v1.5.0-dirty AS pkg-kernel-arm64
+
+#FROM --platform=arm64 ghcr.io/siderolabs/u-boot:${PKGS} AS pkg-u-boot-arm64
+FROM --platform=arm64 127.0.0.1:5000/ubootbash:latest AS pkg-u-boot-arm64
 
-FROM --platform=arm64 ghcr.io/siderolabs/u-boot:${PKGS} AS pkg-u-boot-arm64
 FROM --platform=arm64 ghcr.io/siderolabs/raspberrypi-firmware:${PKGS} AS pkg-raspberrypi-firmware-arm64
 
 # Resolve package images using ${EXTRAS} to be used later in COPY --from=.
